aws-s3-transfer-manager/src/types.rs

```diff
@@ -189,28 +202,84 @@ impl Default for UploadFilter {
 #[derive(Debug)]
 pub struct UploadFilterItem<'a> {
     pub(crate) path: Cow<'a, Path>,
-    pub(crate) metadata: Metadata,
+    pub(crate) symlink_metadata: Metadata,
+    pub(crate) metadata: Option<Metadata>,
```

Comment:
I see it's debatable. I'll add a comment when `metadata` becomes `None`. It is `None` when the upload operation does not follow symbolic links, which is default. If `symlink_metadata.is_symlink()` returns true, both `symlink_metadata.is_file()` and `symlink_metadata.is_dir()` return false, but if the upload operation does not follow symbolic links, that's ok since it does not need to know about the file type of the symlink target.

Along the same reasoning, `metadata` becomes `Some` if the upload operation follows symbolic links.

Reply:
I see it's debatable. I'll add a comment when `metadata` becomes `None`. It is `None` when the upload operation does not follow symbolic links, which is default. If `symlink_metadata.is_symlink()` returns true, both `symlink_metadata.is_file()` and `symlink_metadata.is_dir()` return false, but if the upload operation does not follow symbolic links, `metadata`, the one knows the file type of the symlink target, can be `None` since the upload operation does not need to know about the file type of the symlink target.

Along the same reasoning, `metadata` becomes `Some` if the upload operation follows symbolic links.

Reply:
I see it's debatable. I'll add a comment when `metadata` becomes `None`. It is `None` when the upload operation does not follow symbolic links, which is default. If `symlink_metadata.is_symlink()` returns true, both `symlink_metadata.is_file()` and `symlink_metadata.is_dir()` return false, but if the upload operation does not follow symbolic links, `metadata`, the one knows the file type of the symlink target, can be `None` since the upload operation does not need to know about the file type of the symlink target.

Along the same reasoning, `metadata` becomes `Some` if the upload operation follows symbolic links.

In short, `metadata`, which follows symbolic links, is not suitable for the upload operation's default case where it does not follow symbolic links, so it becomes `Some` only when the upload operation is configured to follow symbolic links.

Reply:
I see it's debatable. I'll add a comment when `metadata` becomes `None`. It is `None` when the upload operation does not follow symbolic links, which is default. If `symlink_metadata.is_symlink()` returns true, both `symlink_metadata.is_file()` and `symlink_metadata.is_dir()` return false, but if the upload operation does not follow symbolic links, `metadata`, the one knows the file type of the symlink target, can be `None` since the upload operation does not need to know about the file type of the symlink target.

Along the same reasoning, `metadata` becomes `Some` if the upload operation follows symbolic links.

In short, `metadata`, which follows symbolic links, is not suitable for the upload operation's default case where it does not follow symbolic links. It is an `Option` because we don't want to pay the additional price of calling `fs::metadata` if we don't need it; we already call `fs::symlink_metadata` to work with upload operation's default case that does not follow symbolic links, and `Metadata` returned by `fs::symlink_metadata` can still answer whether a path entry is a file or a directory if the path entry is not a symbolic link.

Reply:
`symlink_metadata` got removed and `metadata` is no longer an optional in 5a52119.